test/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,79 @@
+// +build upgrade
+
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"testing"
+
+	"go.uber.org/zap"
+	_ "knative.dev/pkg/system/testing"
+	pkgupgrade "knative.dev/pkg/test/upgrade"
+	"knative.dev/serving/test/upgrade/installation"
+)
+
+func TestServingUpgrades(t *testing.T) {
+	c := newUpgradeConfig(t)
+	suite := pkgupgrade.Suite{
+		Tests: pkgupgrade.Tests{
+			PreUpgrade: []pkgupgrade.Operation{
+				ServicePreUpgradeTest(),
+				ServicePreUpgradeAndScaleToZeroTest(),
+				BYORevisionPreUpgradeTest(),
+				InitialScalePreUpgradeTest(),
+			},
+			PostUpgrade: []pkgupgrade.Operation{
+				ServicePostUpgradeTest(),
+				ServicePostUpgradeFromScaleToZeroTest(),
+				BYORevisionPostUpgradeTest(),
+				CreateNewServicePostUpgradeTest(),
+				InitialScalePostUpgradeTest(),
+			},
+			PostDowngrade: []pkgupgrade.Operation{
+				ServicePostDowngradeTest(),
+				CreateNewServicePostDowngradeTest(),
+			},
+			Continual: []pkgupgrade.BackgroundOperation{
+				ProbeTest(),
+				AutoscaleSustainingTest(c.Log),
+				AutoscaleSustainingWithTBCTest(c.Log),
+			},
+		},
+		Installations: pkgupgrade.Installations{
+			Base: []pkgupgrade.Operation{
+				// Do nothing. The initial version is already installed by scripts
+				// together with additional test resources.
```

Comment:
I think we should put here `installation.LatestRelease()` and make scripts do nothing instead. That would give us the ability to integrate that on `knative/operator` and vendor level.

Reply:
The problem is with workarounds that exist for test resources. For example, right now we have disable_chaosduck bash function. This would need to be called separately from Golang -> Bash. And if we have other workaround in the future it will need a separate bash function, and then Golang wrappers. This seems to complicated instead of just using what was setup at the beginning.

Reply:
But I'd like to know other opinions as well.

Reply:
Bot on `knative/operator` on on vendor we will not use serving e2e-upgrade-tests.sh script. Rather, we will have our own entrypoint script there. And then you will need copy & paste serving stuff to it, and manually keep it in sync.

If you moved that to a function then less code will be needed to maintain manually. Take a look how eventing entrypoint script is looking: https://github.com/knative/eventing/blob/6065b38/test/e2e-upgrade-tests.sh#L24-L43 It basically doesn't do anything.

Reply:
Yeah. I know what you mean with respect to re-using this in knative/operator. I've looked at the test setup there. Currently, the installation and setup of test resources are quite different in knative/serving nad knative/operator: installing different images, labeling namespaces differently, waiting for different services/pods. I think the main issue is unifying the code at the level of Bash. I understand you want to enforce that by using these parts in upgrade tests here but this sounds like a problem that should be resolved separately (mainly within Bash). When the Bash functions are fixed/unified they can be re-used by the upgrade framework in knative/operator - when we get to implementing it there. 
However, I wouldn't complicate the Bash scripts and stuff on the knative/serving side right now. IMO, what is there now is the least intrusive change and is in line with the test setup for E2E, and doesn't need wrappers for workarounds.

Reply:
On the second thought we got `installation.LatestRelease()` and `installation.Head()` so we should be good on downstream.